ironfish/src/network/peerNetwork.ts

```diff
@@ -921,37 +1241,99 @@ export class PeerNetwork {
     return new GetCompactBlockResponse(block.toCompactBlock(), message.rpcId)
   }
 
-  private async onNewBlock(peer: Peer, message: NewBlockMessage): Promise<void> {
-    if (!this.enableSyncing) {
+  private async handleNewBlockMessage(peer: Peer, message: NewBlockMessage) {
+    if (!this.shouldProcessNewBlocks()) {
       return
     }
 
-    // Hashes sent by the network are untrusted. Future messages should remove this field.
-    message.block.header.hash = undefined
+    const block = BlockSerde.deserialize(message.block)
+
+    if (await this.alreadyHaveBlock(block.header.hash)) {
+      return
+    }
 
-    const block = message.block
-    const header = BlockHeaderSerde.deserialize(message.block.header)
+    peer.knownBlockHashes.set(block.header.hash, KnownBlockHashesValue.Received)
 
-    peer.knownBlockHashes.set(header.hash, KnownBlockHashesValue.Received)
-    for (const knownPeer of peer.knownPeers.values()) {
-      knownPeer.knownBlockHashes.set(header.hash, KnownBlockHashesValue.Received)
+    // verify the block header
+    const verifyBlockHeaderResult = this.chain.verifier.verifyBlockHeader(block.header)
+    if (!verifyBlockHeaderResult.valid) {
+      this.chain.addInvalid(
+        block.header,
+        verifyBlockHeaderResult.reason ?? VerificationResultReason.ERROR,
+      )
+      return
     }
 
-    try {
-      const result = await this.node.syncer.addNewBlock(peer, block)
-      if (result) {
-        this.broadcastBlock(message)
-      }
+    if (!peer.sequence || block.header.sequence > peer.sequence) {
+      peer.sequence = block.header.sequence
+    }
+
+    // if we don't have the previous block, start syncing
+    const prevHeader = await this.chain.getHeader(block.header.previousBlockHash)
+    if (prevHeader === null) {
+      this.chain.addOrphan(block.header)
+      this.node.syncer.startSync(peer)
       return
-    } catch (error) {
-      this.logger.error(
-        `Error when adding new block ${block.header.sequence} from ${
-          peer.displayName
-        }: ${ErrorUtils.renderError(error, true)}`,
-      )
+    }
+
+    await this.onNewFullBlock(peer, block, prevHeader)
+  }
 
+  private async onNewFullBlock(
+    peer: Peer,
+    block: Block,
+    prevHeader: BlockHeader,
+    nonce?: Buffer,
+  ): Promise<void> {
+    if (!this.shouldProcessNewBlocks()) {
       return
     }
+
+    // Mark that we've assembled a full block in the block fetcher
+    this.blockFetcher.receivedFullBlock(block)
+
+    // Re-gossip the full block
+    const serializedBlock = BlockSerde.serialize(block)
+    const newBlockMessage = new NewBlockMessage(
+      serializedBlock,
+      nonce ?? Buffer.alloc(16, block.header.hash),
+    )
+
+    this.broadcastBlock(newBlockMessage)
```

Comment:
Is there a reason why we broadcast full blocks and compact blocks here but then broadcast the hashes later on?

Reply:
Yeah, it's basically a balance between optimistic propagation of the block and fully validating the block body + merkle root. Similar to ethereum, but we send compact blocks instead: https://github.com/ethereum/devp2p/blob/master/caps/eth.md#block-propagation

Reply:
alright yea that makes sense